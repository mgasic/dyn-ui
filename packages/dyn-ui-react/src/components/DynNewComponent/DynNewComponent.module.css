.root {
  --dyn-new-component-gap: 0;
  --dyn-new-component-padding: 0;
  --dyn-new-component-margin: 0;

  display: flex;
  flex-direction: column;
  align-items: stretch;
  gap: var(--dyn-new-component-gap);
  padding: var(--dyn-new-component-padding);
  margin: var(--dyn-new-component-margin);
  box-sizing: border-box;
}
