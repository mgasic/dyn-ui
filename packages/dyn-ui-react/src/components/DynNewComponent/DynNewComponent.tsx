import { forwardRef } from 'react';
import type { ElementType, ReactElement } from 'react';
import { cn } from '../../utils/classNames';
import {
  DYN_NEW_COMPONENT_DEFAULT_PROPS,
  type DynNewComponentProps,
  type DynNewComponentRef,
  type DynNewComponentSpacingToken,
  type DynNewComponentSpacingValue,
} from './DynNewComponent.types';
import styles from './DynNewComponent.module.css';

type SpacingSlot = 'gap' | 'padding' | 'margin';

const SPACING_TOKENS: Record<Exclude<DynNewComponentSpacingToken, '0' | 'auto'>, string> = {
  xs: '0.25rem',
  sm: '0.5rem',
  md: '1rem',
  lg: '1.5rem',
  xl: '2rem',
  '2xl': '3rem',
};

const resolveSpacingValue = (
  value: DynNewComponentSpacingValue | undefined,
  slot: SpacingSlot
): string | undefined => {
  if (value === undefined) {
    return undefined;
  }

  if (typeof value === 'number') {
    return `${value}px`;
  }

  if (value === 'auto') {
    return slot === 'margin' ? 'auto' : value;
  }

  if (value === '0') {
    return '0';
  }

  if (value in SPACING_TOKENS) {
    const token = SPACING_TOKENS[value as Exclude<DynNewComponentSpacingToken, '0' | 'auto'>];
    return `var(--dyn-spacing-${value}, var(--spacing-${value}, ${token}))`;
  }

  return value;
};

const setSpacingVariable = (
  stylesMap: Record<string, string | number>,
  variable: string,
  value: string | undefined
) => {
  if (value !== undefined) {
    stylesMap[variable] = value;
  }
};

const DynNewComponentInner = <E extends ElementType = 'div'>(
  {
    as,
    gap,
    p,
    m,
    className,
    style,
    children,
    'data-testid': dataTestId = DYN_NEW_COMPONENT_DEFAULT_PROPS['data-testid'],
    ...rest
  }: DynNewComponentProps<E>,
  ref: DynNewComponentRef<E>
) => {
  const Component = (as ?? 'div') as ElementType;

  const spacingStyle: Record<string, string | number> = {
    ...(style ?? {}),
  } as Record<string, string | number>;

  setSpacingVariable(spacingStyle, '--dyn-new-component-gap', resolveSpacingValue(gap, 'gap'));
  setSpacingVariable(spacingStyle, '--dyn-new-component-padding', resolveSpacingValue(p, 'padding'));
  setSpacingVariable(spacingStyle, '--dyn-new-component-margin', resolveSpacingValue(m, 'margin'));

  return (
    <Component
      ref={ref}
      className={cn(styles.root, className)}
      style={spacingStyle}
      data-testid={dataTestId}
      {...rest}
    >
      {children}
    </Component>
  );
};

type DynNewComponentForwardRef = <E extends ElementType = 'div'>(
  props: DynNewComponentProps<E> & { ref?: DynNewComponentRef<E> }
) => ReactElement | null;

const DynNewComponent = forwardRef(DynNewComponentInner as unknown as DynNewComponentForwardRef);

DynNewComponent.displayName = 'DynNewComponent';

export { DynNewComponent };
export default DynNewComponent;
