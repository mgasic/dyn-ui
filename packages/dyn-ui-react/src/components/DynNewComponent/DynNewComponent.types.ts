import type {
  ComponentPropsWithoutRef,
  ComponentRef,
  CSSProperties,
  ElementType,
} from 'react';
import type { BaseComponentProps } from '../../types/theme';

type PolymorphicComponentProps<E extends ElementType, P> = P &
  Omit<ComponentPropsWithoutRef<E>, keyof P | 'as'>;

export type DynNewComponentSpacingToken = '0' | 'xs' | 'sm' | 'md' | 'lg' | 'xl' | '2xl' | 'auto';
export type DynNewComponentSpacingValue =
  | DynNewComponentSpacingToken
  | number
  | string;

export interface DynNewComponentOwnProps extends BaseComponentProps {
  /** Controls the gap between child elements. */
  gap?: DynNewComponentSpacingValue;
  /** Applies padding using spacing tokens, raw CSS values or numbers. */
  p?: DynNewComponentSpacingValue;
  /** Applies margin using spacing tokens, raw CSS values or numbers. */
  m?: DynNewComponentSpacingValue;
  /** Inline style overrides. */
  style?: CSSProperties;
}

export type DynNewComponentBaseProps = DynNewComponentOwnProps;

export type DynNewComponentProps<E extends ElementType = 'div'> =
  PolymorphicComponentProps<E, DynNewComponentBaseProps> & {
    as?: E;
  };

export type DynNewComponentRef<E extends ElementType = 'div'> = ComponentRef<E>;

export interface DynNewComponentDefaultProps {
  'data-testid': string;
}

export const DYN_NEW_COMPONENT_DEFAULT_PROPS: DynNewComponentDefaultProps = {
  'data-testid': 'dyn-new-component',
};
