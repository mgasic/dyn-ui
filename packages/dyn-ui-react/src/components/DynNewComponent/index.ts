export { DynNewComponent } from './DynNewComponent';
export type {
  DynNewComponentProps,
  DynNewComponentRef,
  DynNewComponentSpacingToken,
  DynNewComponentSpacingValue,
} from './DynNewComponent.types';
