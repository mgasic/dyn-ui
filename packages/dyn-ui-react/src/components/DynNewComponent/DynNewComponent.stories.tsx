import type { Meta, StoryObj } from '@storybook/react';
import { DynNewComponent } from './DynNewComponent';
import { ThemeProvider } from '../../theme/ThemeProvider';

const meta: Meta<typeof DynNewComponent> = {
  title: 'Components/DynNewComponent',
  component: DynNewComponent,
  decorators: [Story => (
    <ThemeProvider initialTheme="light">
      <Story />
    </ThemeProvider>
  )],
  argTypes: {
    children: { control: false },
  },
};

export default meta;

type Story = StoryObj<typeof DynNewComponent>;

export const WrapperExample: Story = {
  args: {
    gap: 'md',
    p: 'lg',
    m: 'sm',
    as: 'section',
    children: (
      <>
        <div style={{ background: 'var(--dyn-color-surface, #f5f5f5)', padding: '0.5rem' }}>Header content</div>
        <div style={{ background: 'var(--dyn-color-surface, #f5f5f5)', padding: '0.5rem' }}>Body content</div>
        <div style={{ background: 'var(--dyn-color-surface, #f5f5f5)', padding: '0.5rem' }}>Footer content</div>
      </>
    ),
  },
  parameters: {
    layout: 'centered',
  },
};
