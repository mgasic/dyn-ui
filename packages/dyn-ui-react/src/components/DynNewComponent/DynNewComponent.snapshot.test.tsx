import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { DynNewComponent } from './DynNewComponent';

describe('DynNewComponent snapshots', () => {
  it('matches the default wrapper snapshot', () => {
    const { container } = render(
      <DynNewComponent gap="md" p="lg" m="sm">
        <div>First block</div>
        <div>Second block</div>
      </DynNewComponent>
    );

    expect(container).toMatchSnapshot();
  });
});
