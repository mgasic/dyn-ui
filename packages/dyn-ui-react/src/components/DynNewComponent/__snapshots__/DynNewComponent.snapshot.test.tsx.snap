// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DynNewComponent snapshots > matches the default wrapper snapshot 1`] = `
<div>
  <div
    class="_root_052305"
    data-testid="dyn-new-component"
    style="--dyn-new-component-gap: var(--dyn-spacing-md, var(--spacing-md, 1rem)); --dyn-new-component-padding: var(--dyn-spacing-lg, var(--spacing-lg, 1.5rem)); --dyn-new-component-margin: var(--dyn-spacing-sm, var(--spacing-sm, 0.5rem));"
  >
    <div>
      First block
    </div>
    <div>
      Second block
    </div>
  </div>
</div>
`;
